File: standings.py
"""
Módulo para la clase Standings (Tabla de Posiciones).

Gestiona la tabla de posiciones de una categoría.
"""

from typing import List, Dict
from team import Team


class Standings:
    """Gestiona la tabla de posiciones de una categoría."""

    def __init__(self, category: str, points_per_win: int = 2, points_per_loss: int = 0):
        """
        Inicializa la tabla de posiciones.

        Args:
            category: Nombre de la categoría
            points_per_win: Puntos otorgados por victoria
            points_per_loss: Puntos otorgados por derrota
        """
        self.category = category
        self.points_per_win = points_per_win
        self.points_per_loss = points_per_loss
        self.teams: Dict[str, Team] = {}

    def add_team(self, team: Team):
        """
        Agrega un equipo a la tabla de posiciones.

        Args:
            team: Instancia de Team
        """
        self.teams[team.name] = team

    def update_standings(self):
        """Recalcula todas las posiciones de la tabla."""
        for team in self.teams.values():
            team.calculate_points(self.points_per_win, self.points_per_loss)

    def get_sorted_standings(self) -> List[Team]:
        """
        Obtiene la tabla de posiciones ordenada.

        Ordena por:
        1. Puntos totales (descendente)
        2. Diferencia de puntos (descendente)
        3. Puntos a favor (descendente)
        4. Nombre (alfabético)

        Returns:
            Lista de equipos ordenados
        """
        self.update_standings()
        sorted_teams = sorted(
            self.teams.values(),
            key=lambda t: (-t.points, -t.get_difference(), -t.pf, t.name)
        )
        return sorted_teams

    def get_team(self, team_name: str) -> Team:
        """
        Obtiene un equipo por su nombre.

        Args:
            team_name: Nombre del equipo

        Returns:
            Instancia de Team
        """
        return self.teams.get(team_name)

    def apply_penalty(self, team_name: str, points: int):
        """
        Aplica una multa o bonificación de puntos a un equipo.

        Args:
            team_name: Nombre del equipo
            points: Puntos a agregar (positivo) o restar (negativo)
        """
        if team_name in self.teams:
            self.teams[team_name].add_penalty(points)
            self.update_standings()

    def to_dict(self) -> dict:
        """
        Convierte la tabla de posiciones a un diccionario.

        Returns:
            Diccionario con la tabla de posiciones
        """
        sorted_teams = self.get_sorted_standings()
        return {
            "category": self.category,
            "points_per_win": self.points_per_win,
            "points_per_loss": self.points_per_loss,
            "standings": [team.to_dict() for team in sorted_teams]
        }

    def __repr__(self) -> str:
        return f"Standings(category='{self.category}', teams={len(self.teams)})"

File: src/models/Standings.ts
import { Team } from './Team';

/**
 * Gestiona la tabla de posiciones de una categoría.
 */
export class Standings {
  category: string;
  pointsPerWin: number;
  pointsPerLoss: number;
  teams: Map<string, Team> = new Map();

  constructor(category: string, pointsPerWin: number = 2, pointsPerLoss: number = 0) {
    this.category = category;
    this.pointsPerWin = pointsPerWin;
    this.pointsPerLoss = pointsPerLoss;
  }

  /**
   * Agrega un equipo a la tabla de posiciones.
   */
  addTeam(team: Team): void {
    this.teams.set(team.name, team);
  }

  /**
   * Recalcula todas las posiciones de la tabla.
   */
  updateStandings(): void {
    for (const team of this.teams.values()) {
      team.calculatePoints(this.pointsPerWin, this.pointsPerLoss);
    }
  }

  /**
   * Obtiene la tabla de posiciones ordenada.
   * Ordena por: puntos totales, diferencia, puntos a favor, nombre.
   */
  getSortedStandings(): Team[] {
    this.updateStandings();
    const sortedTeams = Array.from(this.teams.values()).sort((a, b) => {
      // 1. Por puntos (descendente)
      if (b.points !== a.points) {
        return b.points - a.points;
      }
      // 2. Por diferencia (descendente)
      const diffA = a.getDifference();
      const diffB = b.getDifference();
      if (diffB !== diffA) {
        return diffB - diffA;
      }
      // 3. Por puntos a favor (descendente)
      if (b.pf !== a.pf) {
        return b.pf - a.pf;
      }
      // 4. Por nombre (alfabético)
      return a.name.localeCompare(b.name);
    });
    return sortedTeams;
  }

  /**
   * Obtiene un equipo por su nombre.
   */
  getTeam(teamName: string): Team | undefined {
    return this.teams.get(teamName);
  }

  /**
   * Aplica una multa o bonificación de puntos a un equipo.
   */
  applyPenalty(teamName: string, points: number): void {
    const team = this.teams.get(teamName);
    if (team) {
      team.addPenalty(points);
      this.updateStandings();
    }
  }

  /**
   * Convierte la tabla de posiciones a un objeto JSON.
   */
  toDict(): Record<string, any> {
    const sortedTeams = this.getSortedStandings();
    return {
      category: this.category,
      points_per_win: this.pointsPerWin,
      points_per_loss: this.pointsPerLoss,
      standings: sortedTeams.map(team => team.toDict())
    };
  }

  toString(): string {
    return `Standings(category='${this.category}', teams=${this.teams.size})`;
  }
}




